acceptancetests/assess_persistent_storage.py

```diff
@@ -1,139 +1,428 @@
-                'Expected token "{}" not found in stored results.'.format(
-                    expected_name))
-        if stored_value != expected_value:
-            raise JujuAssertionError(
-                'Token values do not match. Expected: {} got: {}'.format(
-                    expected_value, stored_value))
 
-
-def get_stored_token_content(client):
-    """Retrieve token values from file stored on the unit.
-
-    Token values are retrieved from the storage units and logged into a file on
-    the application unit.
-
-    :param client: ModelClient object to query.
-    :return: Dict containing 'token name' -> 'token value'.
-    """
-    try:
-        # TODO: Need to pass in application/unit
-        contents = client.get_juju_output(
-            'ssh', ('dummy-storage/0', 'cat', '/tmp/status'))
-    except CalledProcessError as e:
-        raise JujuAssertionError('Failed to read token file: {}'.format(e))
-
-    return {token_name: token_value for token_name, token_value in (
-        line.split(':', 1) for line in contents.splitlines() if line != '')}
+    single_blk_id = \
+        assess_charm_removal_single_block_and_filesystem_storage(client)
+    charm_name = 'dummy-storage'
+    charm_path = local_charm_path(
+        charm=charm_name, juju_ver=client.version)
+    # Run juju deploy dummy-storage --attach-storage <single_blk_id>
+    # Note: --attach-storage has the issue of Bug 1704105 too:
+    # https://bugs.launchpad.net/juju/+bug/1704105
+    client.get_juju_output(
+        'deploy', charm_path,
+        '--attach-storage', single_blk_id,
+        include_e=False, merge_stderr=True)
+    client.wait_for_started()
+    client.wait_for_workloads()
+    assert_app_status(client, charm_name=charm_name, expected='active')
+    storage_list,\
+    storage_type,\
+    persistent_setting,\
+    pool_storage,\
+    pool_setting,\
+    storage_status = get_storage_property(client, storage_id=single_blk_id)
+    assert_storage_number(storage_list=storage_list, expected=1)
+    assert_single_blk_existence(
+        storage_list=storage_list, storage_id=single_blk_id)
+    assert_persistent_setting(
+        storage_id=single_blk_id, found=persistent_setting, expected=True)
+    assert_storage_status(
+        found_id=pool_storage, expected_id=single_blk_id,
+        found_status=storage_status, expected_status='attached')
+    # Run juju remove-application dummy-storage
+    client.get_juju_output('remove-application', charm_name, include_e=False)
+    time.sleep(90)
+    # persistent storage single_blk_id should remain after remove-application
+    assert_single_blk_existence(
+        storage_list=storage_list, storage_id=single_blk_id)
+    # Run juju remove-storage <single_blk_id>
+    # Note: remove-storage has the issue of Bug 1704105 too:
+    # https://bugs.launchpad.net/juju/+bug/1704105
+    client.get_juju_output(
+        'remove-storage', single_blk_id, include_e=False, merge_stderr=True)
+    time.sleep(90)
```

Comment:
I have also considered this scenario, which actually is equivalent to time.sleep().
Say I create a function wait_for_storage_removal(), inside I put remove_storage, get_storage or whatever command, then set a sleep interval for, say 30 seconds. No matter what I do, I need to set a total sleep time out limit, in this case it's 90 seconds. So the only difference is with function wait_for_storage_removal(), it runs the command and status check 2 more times, and if the command can't give an expected return within 60 seconds, it will attempt to run the third time, which would become exactly the same as time.sleep(90) in terms of waiting time. The only benefit would be the command returns expected result within 60 seconds or even within 30 seconds (highly unlikely), and we can save that 30+ seconds. Also, I can't set the time interval too short, e.g.: 15 seconds, which may cause the issue I mentioned in the previous comments.
To avoid all this fuss, I decided to put the simplest way there. As I mentioned, the "watcher" mechanism in Juju seems to be the right path to go with but I'm foreign to that area, will need time to learn.